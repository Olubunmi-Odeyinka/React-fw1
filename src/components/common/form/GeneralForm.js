import React from 'react';
import PropTypes from 'prop-types';
import {Form} from 'formik';
import { NavLink} from 'react-router-dom';
import TextInput from './input/TextInput';
import SelectInput from './input/SelectInput';
import _ from  'lodash';

export default class GeneralForm extends React.Component {
  // constructor(props, context){
  //   super(props, context);
  //
  //   this.renderField = this.renderField.bind(this);
  // }

  renderField =(field)=>{
   // const fieldHelper = this.props.fields[field];
    const fieldConfig = this.props.fields[field];
    switch(fieldConfig.type){
      case 'select':
        return (
          <SelectInput
            key={field}
            name={field}
            focus={fieldConfig.focus}
            label={fieldConfig.label  || field}
            error={this.props.errors[field] || null}
            touched={this.props.touched[field] || null}
            options={this.props.lookUps[field] || null}
            readOnly={this.props.formState.formFieldReadOnly}/>
        );
      default:
        return (
          <TextInput
            key={field}
            name={field}
            focus={fieldConfig.focus}
            type={fieldConfig.type}
            label={fieldConfig.label || field}
            touched={this.props.touched[field] || null}
            error={this.props.errors[field] || null}
            readOnly={this.props.formState.formFieldReadOnly}/>
        );
    }
  }

  render =()=> {
    return (
      <div className="container">
        <div className="row">
          <h1>{this.props.formState.formHeader}</h1>
          <NavLink className="btn btn-default pull-right" to={this.props.formState.listUrl}>Back to List</NavLink>
        </div>
        <Form>
          {_.map(this.props.fields, (val, key)=> this.renderField(key))}

          {this.props.formState.showSavingButton ?
          (<input
            type="submit"
            disabled={this.props.saving}
            value={this.props.saving ? this.props.formState.savingString[0]+'...' : this.props.formState.savingString[1]}
            className={"btn "+ this.props.formState.savingString[2] || "btn-info"}
            onClick={this.props.onSave}/>) : null}
        </Form>
      </div>
    );
  }
}


// GeneralForm.propTypes = {
//   formHeader: PropTypes.string.isRequired,
//   formModule: PropTypes.object.isRequired,
//   fields: PropTypes.object.isRequired,
//   lookUps: PropTypes.object,
//   onSave: PropTypes.func.isRequired,
//   onChange: PropTypes.func.isRequired,
//   saving: PropTypes.bool,
//   errors: PropTypes.object
// };

//export default GeneralForm;
