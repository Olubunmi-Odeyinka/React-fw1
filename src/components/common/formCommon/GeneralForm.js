import React from 'react';
import PropTypes from 'prop-types';
import TextInput from './inputCommon/TextInput';
import SelectInput from './inputCommon/SelectInput';
import _ from  'lodash';

export default class GeneralForm extends React.Component {
  constructor(props, context){
    super(props, context);

    this.renderField = this.renderField.bind(this);
  }

  renderField(field){
   // const fieldHelper = this.props.fields[field];
    const fieldConfig = this.props.fields[field];
    switch(fieldConfig.type){
      case 'select':
        return (
          <SelectInput
            key={field}
            name={field}
            label={fieldConfig.label}
            value={this.props.formModule[field]}
            onChange={this.props.onChange}
            error={this.props.errors[field] || null}
            defaultOption={this.props.fields[field].defaultOption || null}
            options={this.props.lookUps[field] || null} />
        );
      default:
        return (
          <TextInput
            key={field}
            name={field}
            type={fieldConfig.type}
            label={fieldConfig.label}
            value={this.props.formModule[field]}
            onChange={this.props.onChange}
            error={this.props.errors[field] || null} />
        );
    }
  }

  render() {
    return (
      <div className="container">
        <form>
          <h1>{this.props.formHeader}</h1>

          {_.map(this.props.fields, (val, key)=> this.renderField(key))}

          <input
            type="submit"
            disabled={this.props.saving}
            value={this.props.saving ? 'Saving...' : 'Save'}
            className="btn btn-primary"
            onClick={this.props.onSave}/>
        </form>
      </div>
    );
  }
}

GeneralForm.propTypes = {
  formHeader: PropTypes.string.isRequired,
  formModule: PropTypes.object.isRequired,
  fields: PropTypes.object.isRequired,
  lookUps: PropTypes.object,
  onSave: PropTypes.func.isRequired,
  onChange: PropTypes.func.isRequired,
  saving: PropTypes.bool,
  errors: PropTypes.object
};

//export default GeneralForm;
