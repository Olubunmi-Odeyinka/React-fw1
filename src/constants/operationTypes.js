
export const Create_URL_String = 'create';
export const Modify_URL_String = 'modify';
export const Delete_URL_String = 'delete';
