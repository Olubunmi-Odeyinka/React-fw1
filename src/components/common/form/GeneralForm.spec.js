import React from 'react';
import { shallow } from 'enzyme';
import GeneralForm from './GeneralForm';
import * as operation from '../../../constants/operationTypes';

let state, props, fieldsDefinition;

beforeEach(() => {
  state = {
    formHeader: "Manage Books",
    formFieldReadOnly: false,
    urlOpeartion: '',
    listUrl:'/books',
    showSavingButton: true,
    savingString:['saving', 'save', 'btn-success']
  };

  fieldsDefinition = {
    name: {
      type: 'text',
      label: 'Name'
    },
    author: {
      type: 'text',
      label: 'Author',
      focus: true
    },
    category: {
      type: 'select',
      label: 'Category'
    },
    page: {
      type: 'number',
      label: 'Book\'s Pages'
    }
  };

  props = {
    errors: {},
    fields: fieldsDefinition,
    touched: {},
    lookUps: {},
    formState: state,
    saving: false,
    onSave: true
  };
});

test('b', ()=>{

  const wrapper = shallow(<GeneralForm  {...props}/>);
  expect(wrapper).toMatchSnapshot();
});

