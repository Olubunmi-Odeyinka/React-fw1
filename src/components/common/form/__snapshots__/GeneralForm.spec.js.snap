// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`b 1`] = `
ShallowWrapper {
  "length": 1,
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <GeneralForm
    errors={Object {}}
    fields={
        Object {
            "author": Object {
              "focus": true,
              "label": "Author",
              "type": "text",
            },
            "category": Object {
              "label": "Category",
              "type": "select",
            },
            "name": Object {
              "label": "Name",
              "type": "text",
            },
            "page": Object {
              "label": "Book's Pages",
              "type": "number",
            },
          }
    }
    formState={
        Object {
            "formFieldReadOnly": false,
            "formHeader": "Manage Books",
            "listUrl": "/books",
            "savingString": Array [
              "saving",
              "save",
              "btn-success",
            ],
            "showSavingButton": true,
            "urlOpeartion": "",
          }
    }
    lookUps={Object {}}
    onSave={true}
    saving={false}
    touched={Object {}}
/>,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <div
          className="row"
>
          <h1>
                    Manage Books
          </h1>
          <NavLink
                    activeClassName="active"
                    ariaCurrent="true"
                    className="btn btn-default pull-right"
                    to="/books"
          >
                    Back to List
          </NavLink>
</div>,
        <Form>
          <TextInput
                    error={null}
                    focus={undefined}
                    label="Name"
                    name="name"
                    readOnly={false}
                    touched={null}
                    type="text"
          />
          <TextInput
                    error={null}
                    focus={true}
                    label="Author"
                    name="author"
                    readOnly={false}
                    touched={null}
                    type="text"
          />
          <SelectInput
                    error={null}
                    focus={undefined}
                    label="Category"
                    name="category"
                    options={null}
                    readOnly={false}
                    touched={null}
          />
          <TextInput
                    error={null}
                    focus={undefined}
                    label="Book's Pages"
                    name="page"
                    readOnly={false}
                    touched={null}
                    type="number"
          />
          <input
                    className="btn btn-success"
                    disabled={false}
                    onClick={true}
                    type="submit"
                    value="save"
          />
</Form>,
      ],
      "className": "container",
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": Array [
            <h1>
              Manage Books
</h1>,
            <NavLink
              activeClassName="active"
              ariaCurrent="true"
              className="btn btn-default pull-right"
              to="/books"
>
              Back to List
</NavLink>,
          ],
          "className": "row",
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "children": "Manage Books",
            },
            "ref": null,
            "rendered": "Manage Books",
            "type": "h1",
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "function",
            "props": Object {
              "activeClassName": "active",
              "ariaCurrent": "true",
              "children": "Back to List",
              "className": "btn btn-default pull-right",
              "to": "/books",
            },
            "ref": null,
            "rendered": "Back to List",
            "type": [Function],
          },
        ],
        "type": "div",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "children": Array [
            Array [
              <TextInput
                error={null}
                focus={undefined}
                label="Name"
                name="name"
                readOnly={false}
                touched={null}
                type="text"
/>,
              <TextInput
                error={null}
                focus={true}
                label="Author"
                name="author"
                readOnly={false}
                touched={null}
                type="text"
/>,
              <SelectInput
                error={null}
                focus={undefined}
                label="Category"
                name="category"
                options={null}
                readOnly={false}
                touched={null}
/>,
              <TextInput
                error={null}
                focus={undefined}
                label="Book's Pages"
                name="page"
                readOnly={false}
                touched={null}
                type="number"
/>,
            ],
            <input
              className="btn btn-success"
              disabled={false}
              onClick={true}
              type="submit"
              value="save"
/>,
          ],
        },
        "ref": null,
        "rendered": Array [
          Object {
            "instance": null,
            "key": "name",
            "nodeType": "function",
            "props": Object {
              "error": null,
              "focus": undefined,
              "label": "Name",
              "name": "name",
              "readOnly": false,
              "touched": null,
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "author",
            "nodeType": "function",
            "props": Object {
              "error": null,
              "focus": true,
              "label": "Author",
              "name": "author",
              "readOnly": false,
              "touched": null,
              "type": "text",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "category",
            "nodeType": "function",
            "props": Object {
              "error": null,
              "focus": undefined,
              "label": "Category",
              "name": "category",
              "options": null,
              "readOnly": false,
              "touched": null,
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": "page",
            "nodeType": "function",
            "props": Object {
              "error": null,
              "focus": undefined,
              "label": "Book's Pages",
              "name": "page",
              "readOnly": false,
              "touched": null,
              "type": "number",
            },
            "ref": null,
            "rendered": null,
            "type": [Function],
          },
          Object {
            "instance": null,
            "key": undefined,
            "nodeType": "host",
            "props": Object {
              "className": "btn btn-success",
              "disabled": false,
              "onClick": true,
              "type": "submit",
              "value": "save",
            },
            "ref": null,
            "rendered": null,
            "type": "input",
          },
        ],
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <div
            className="row"
>
            <h1>
                        Manage Books
            </h1>
            <NavLink
                        activeClassName="active"
                        ariaCurrent="true"
                        className="btn btn-default pull-right"
                        to="/books"
            >
                        Back to List
            </NavLink>
</div>,
          <Form>
            <TextInput
                        error={null}
                        focus={undefined}
                        label="Name"
                        name="name"
                        readOnly={false}
                        touched={null}
                        type="text"
            />
            <TextInput
                        error={null}
                        focus={true}
                        label="Author"
                        name="author"
                        readOnly={false}
                        touched={null}
                        type="text"
            />
            <SelectInput
                        error={null}
                        focus={undefined}
                        label="Category"
                        name="category"
                        options={null}
                        readOnly={false}
                        touched={null}
            />
            <TextInput
                        error={null}
                        focus={undefined}
                        label="Book's Pages"
                        name="page"
                        readOnly={false}
                        touched={null}
                        type="number"
            />
            <input
                        className="btn btn-success"
                        disabled={false}
                        onClick={true}
                        type="submit"
                        value="save"
            />
</Form>,
        ],
        "className": "container",
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": Array [
              <h1>
                Manage Books
</h1>,
              <NavLink
                activeClassName="active"
                ariaCurrent="true"
                className="btn btn-default pull-right"
                to="/books"
>
                Back to List
</NavLink>,
            ],
            "className": "row",
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "children": "Manage Books",
              },
              "ref": null,
              "rendered": "Manage Books",
              "type": "h1",
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "function",
              "props": Object {
                "activeClassName": "active",
                "ariaCurrent": "true",
                "children": "Back to List",
                "className": "btn btn-default pull-right",
                "to": "/books",
              },
              "ref": null,
              "rendered": "Back to List",
              "type": [Function],
            },
          ],
          "type": "div",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "children": Array [
              Array [
                <TextInput
                  error={null}
                  focus={undefined}
                  label="Name"
                  name="name"
                  readOnly={false}
                  touched={null}
                  type="text"
/>,
                <TextInput
                  error={null}
                  focus={true}
                  label="Author"
                  name="author"
                  readOnly={false}
                  touched={null}
                  type="text"
/>,
                <SelectInput
                  error={null}
                  focus={undefined}
                  label="Category"
                  name="category"
                  options={null}
                  readOnly={false}
                  touched={null}
/>,
                <TextInput
                  error={null}
                  focus={undefined}
                  label="Book's Pages"
                  name="page"
                  readOnly={false}
                  touched={null}
                  type="number"
/>,
              ],
              <input
                className="btn btn-success"
                disabled={false}
                onClick={true}
                type="submit"
                value="save"
/>,
            ],
          },
          "ref": null,
          "rendered": Array [
            Object {
              "instance": null,
              "key": "name",
              "nodeType": "function",
              "props": Object {
                "error": null,
                "focus": undefined,
                "label": "Name",
                "name": "name",
                "readOnly": false,
                "touched": null,
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "author",
              "nodeType": "function",
              "props": Object {
                "error": null,
                "focus": true,
                "label": "Author",
                "name": "author",
                "readOnly": false,
                "touched": null,
                "type": "text",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "category",
              "nodeType": "function",
              "props": Object {
                "error": null,
                "focus": undefined,
                "label": "Category",
                "name": "category",
                "options": null,
                "readOnly": false,
                "touched": null,
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": "page",
              "nodeType": "function",
              "props": Object {
                "error": null,
                "focus": undefined,
                "label": "Book's Pages",
                "name": "page",
                "readOnly": false,
                "touched": null,
                "type": "number",
              },
              "ref": null,
              "rendered": null,
              "type": [Function],
            },
            Object {
              "instance": null,
              "key": undefined,
              "nodeType": "host",
              "props": Object {
                "className": "btn btn-success",
                "disabled": false,
                "onClick": true,
                "type": "submit",
                "value": "save",
              },
              "ref": null,
              "rendered": null,
              "type": "input",
            },
          ],
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
      },
    },
  },
}
`;
